### DISCUSSION TITLE
0.5.0 pre

### DISCUSSION THREAD

**Discussion by Vortex2Oblivion** [Vortex2Oblivion/LeatherEngine-Extended-Support#47] [Announcements]
## 0.5.0pre

The full release of 0.5.0 has been in development for a while now and with mainly only bugfixes left its time for a stable prerelease.

In case you haven't been keeping up here's whats changed since 0.4.2

### Added

- Weekend 1
- Support for vslice chart format.
- Zoom in the chart editor.
- Lots of new functions and variables in the Lua API. See the wiki!
- `singDuration` property to character jsons to control their sing duration.
- Flixel Splash Screen option to toggle the HaxeFlixel splash screen that shows up when the game starts.
- Icons now bump to the song playing when you play it with the space bar in freeplay.
- When using the default time bar style the time bar color changes depending on what character is currently in focus.
- Score text changed to use the format `< Score:$score ~ Misses:$misses ~ Accuracy:$accuracy% ~ $comboRating ~ $rating >
- Animation Debug now has a toggleable stage and a character position dot to help with offsets.
- You can now switch what side you are using to edit offsets on in Animation Debug.
- `Mute Vocals` option in chart editor.
- Persistent Cached Data option (stops clearing asset caches).
- Vram Sprites option (loads textures into gpu)
- Skip Results Screen Options.
- You can now zoom into the Animation Debug using the mouse.
- HScript support!
- Global scripts will be run on any mod in data/scripts/global
- Local scripts will be run on the current mod in data/scripts/local
- You can put any .lua or .hx file into a song folder to have it run (like psych engine)
- Discord Rich Presence can now be changed per mod.
- Runtime shader support
- Change Keycount events
- Copy/Paste events in charter
- Cool new modchart features via modcharting tools haxelib
- Modcharting tools can be used in lua scripts, hscript, and the modchart editor (hit 9 midsong)
- Lil' Buddies in charter
- Multiatlas support in characters via `extraSheets` property
- FlxAnimate support (Texture atlas)
- `mainCharacterID` property in group character jsons
- Characters can now have an hscript attached.
- Added `script` type to cutscenes.
- Added `introSound` property to cutscenes.
- Notes can now be skewed
- Notes can now carry a hscript or lua script.
- Color quants option.
- Stages can now have a hscript.
- Stages have a HSV shader attached, used in voiid chronicles.
- `backgroundColor` property in stages.
- `imageDirectory` property in stages.
- `dances` property in stage objects
- The mod system has been changed to allow for more customization and less conflicts. The currently selected mod will take priority over all other mods.
- The base game assets are now in a mod, this mod will take priority 2nd to the currently selected mod.
- All states/and substates can now carry a script.
- Added developer mode option. Shows ingame logs. Press F3 when enabled.
- Different breakfast themes per character.
- Ingame mod creator gui.
- Added support for hud elements to be moved.
- The ability to write custom states and substates via hscript

### Fixed

- Bug with charting state (specifically `specialAudioName` not working in it).
- Bug where you can go to negative song position in the charter.
- Bug where the ui skin menu opens when opening the `Note Options` section in the options menu.
- Bug where the game would crash in the pause menu when `Prototype Title Screen` is on.
- Bug where the game would crash when opening the ui skin menu with an invalid ui skin in your save data.
- Bug with anything that would normally open the 'application error' window that crashed on Linux.
- Bug where pause menu music wouldn't disappear after closing the menu.
- Bug where the game crashes when trying to load a song or week set that does not exist in story mode.
- Animations now play when character will idle is true like how they would in psych (aka animation fully plays then once it stops the character idles).
- That one bug where characters that dance left and right would be slower than needed at high bpms.
- Difficulty specific audio not playing in freeplay (Linux).
- `crochet` now links to `Conductor.crochet` instead of `Conductor.stepCrochet` in lua (WARNING: This fix may accidentally break some obsecure scripts, to fix them use the new `stepCrochet` variable instead).
- Some alphabet font characters having the wrong name internally, causing them to not appear correctly in-game.
- Not being able to set note or strum angles in lua.
- Pause Menu now uses it's own camera so that it isn't affected by the game's ui zoom.
- Event Luas now have the `create()` function called on them just like the rest of the game's lua.
- Crashes with sounds not having a `_transform` value while trying to use it.
- Memory Counter reports correct values.
- Note splashes should now be antialiased properly.
- Lots of new built in events.
- Skip time option when charting.
- Change playback speed when charting,
- Bug where charts in psych extra keys would display the wrong key count.
- All cached assets should try and clear when changing states to hopefully help on memory leaks.

### Changed

- The option `Import Old Scores` is now in `Misc`.
- Optimized Note Splash spawning.
- Optimized Score text and Info text in PlayState.
- Pause menu music is now cached at the start of each song to prevent lag when opening the menu.
- Freeplay no longer has < > when the selected song has one difficulty.
- Info text now goes onto the time bar in the leather engine time bar style.
- Stage sprites that use the `beatHit` animation no longer are forced to play it every beat (if it's already playing it won't play until it's not).
- Tutorial's camera zoom now acts like it does in base game\* (not exactly the same, but close enough, and a less janky way as well).
- Some miscellaneous ui changes in-game.
- Revamped Animation Debug Menu.
- Results screen note graph now uses one texture for much better performance.
- Hitting 8 midsong will now bring you to animation debug
- Better tankman offsets
- Optimized all spritesheets
- Results screen now matches the one from vslice
- Softcoded all stages and characters
- Added support for voiid chronicles lua functions.
- HSV note shader to RGB note shader (may break some old ui skins)
- The time bar is now a group
- Updated polymod to 1.8.0
- All menu backgrounds are now recolored ingame.
- Story mode and freeplay will no longer crash when no weeks or songs are present.
- Window title is now named after the current mod rather than a text file.

### Removed
 - LeatherLogoBumpin
 - Unused spritesheets
 - Support for shaggy charts (The ones made in the ke 1.2 3 year old mod that nobody has touched in forever)
 - Replays support (nobody used them and they caused a lot of overhead)

And anything else im forgetting

**Full Changelog**: https://github.com/Vortex2Oblivion/LeatherEngine/commits/0.5.0pre

<hr /><em>This discussion was created from the release <a href='https://github.com/Vortex2Oblivion/LeatherEngine/releases/tag/0.5.0pre'>0.5.0 pre</a>.</em>

**Comment by ghost** [comment]
we're so barack
